src/vdaf/mastic.rs

```diff
@@ -0,0 +1,651 @@
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    algorithm_id: u32,
+    szk: Szk<T, P, SEED_SIZE>,
+    pub(crate) vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+    /// The length of the private attribute associated with any input.
+    pub(crate) bits: usize,
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Creates a new instance of Mastic, with a specific attribute length and weight type.
+    pub fn new(
+        algorithm_id: u32,
+        szk: Szk<T, P, SEED_SIZE>,
+        vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+        bits: usize,
+    ) -> Self {
+        Self {
+            algorithm_id,
+            szk,
+            vidpf,
+            bits,
+        }
+    }
+}
+
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation and a set of prefixes to evaluate at that level.
+#[derive(Clone, Debug)]
+pub struct MasticAggregationParam {
+    /// aggregation parameter inherited from [`Poplar1`]: contains the level (attribute length) and a vector of attributes (IdpfInputs)
+    poplar_param: Poplar1AggregationParam,
+    /// Flag indicating whether the VIDPF weight needs to be validated using SZK.
+    root_check_flag: bool,
+}
+
+impl Encode for MasticAggregationParam {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.poplar_param.encode(bytes)?;
+        let root_check = if self.root_check_flag { 1u8 } else { 0u8 };
+        root_check.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.poplar_param.encoded_len()? + 1usize)
+    }
+}
+
+impl Decode for MasticAggregationParam {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let poplar_param = Poplar1AggregationParam::decode(bytes)?;
+        let root_check = u8::decode(bytes)?;
+        let root_check_flag = root_check == 0;
+        Ok(Self {
+            poplar_param,
+            root_check_flag,
+        })
+    }
+}
+
+/// Mastic public share
+///
+/// Contains broadcast information shared between parties to support VIDPF correctness.
+pub type MasticPublicShare<V> = VidpfPublicShare<V>;
+
+impl<T, P, const SEED_SIZE: usize> ParameterizedDecode<Mastic<T, P, SEED_SIZE>>
+    for MasticPublicShare<VidpfWeight<T::Field>>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        mastic: &Mastic<T, P, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        MasticPublicShare::<VidpfWeight<T::Field>>::decode_with_param(
+            &(mastic.bits, mastic.vidpf.weight_parameter),
+            bytes,
+        )
+    }
+}
+
+/// Mastic input share
+///
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
```

Comment:
Why is this variable called "proofs_share"? We use this name in Prio3 to denote the fact that there may be multiple, independent proofs generated and verified. Is that also the case for Mastic?

Reply:
Nope, fixed now.